import sys,os,tempfile
import subprocess
import re
from PySide6.QtWidgets import (QApplication, QMainWindow, QFileDialog, QMessageBox,QProgressDialog)
from PySide6.QtCore import Qt, QTimer
from PySide6.QtGui import QPixmap, QCloseEvent
from PySide6.QtCore import Qt

from core.utils import get_ext_path, get_app_dir, init_config_files
from core.worker import FFmpegWorker
from core.engine import get_video_duration, probe_video_info
from ui.ui_main_window import Ui_MainWindow

class FFmpegGUI(QMainWindow, Ui_MainWindow):
    def __init__(self):
        super().__init__()
        # ==========================================
        # 1. é­”æ³•å¯åŠ¨ï¼šä¸€è¡Œä»£ç åŠ è½½æ‰€æœ‰ç”Ÿæˆçš„ç•Œé¢å…ƒç´ 
        # ==========================================
        self.setupUi(self)

        # ==========================================
        # 2. ä¿ç•™åŸæœ‰çš„æ ¸å¿ƒåˆå§‹åŒ–é€»è¾‘ï¼šç¡¬ä»¶è‡ªæ£€ä¸åŠ¨æ€é¢„è®¾
        # ==========================================
        self.available_v_encoders = self.probe_hardware_encoders()
        self.load_dynamic_presets()

        # ç»™ UI ä¸­ç©ºç™½çš„ä¸‹æ‹‰èœå•åŠ¨æ€å¡å…¥æ•°æ®
        self.cb_v_encoder.addItems(self.available_v_encoders)
        self.combo_preset.addItems(list(self.preset_configs.keys()))
        self.toggle_custom_tab(self.combo_preset.currentText())
        
        # === ä»å¤–éƒ¨ YAML ç»Ÿä¸€åŠ è½½å¹¶æ³¨å…¥æ‰€æœ‰æ‚¬æµ®æç¤º ===
        self.load_tooltips()

        # ==========================================
        # 4. é€å­—ä¿ç•™ï¼šåŸæœ‰çš„æ‰€æœ‰äº‹ä»¶ç»‘å®šé€»è¾‘
        # ==========================================
        # é¢„è®¾è”åŠ¨ä¸æ¨¡å¼åˆ‡æ¢
        self.combo_preset.currentTextChanged.connect(self.toggle_custom_tab)
        self.cb_v_rc.currentTextChanged.connect(self.update_slider_range)
        self.sld_v_value.valueChanged.connect(self.update_slider_label)
        self.cb_format.currentTextChanged.connect(self.change_output_extension)
        
        # æŒ‰é’®ç»‘å®š
        self.btn_pause.clicked.connect(self.toggle_pause)
        self.btn_stop.clicked.connect(self.stop_encoding)
        self.btn_start.clicked.connect(self.start_encoding)
        self.btn_input.clicked.connect(self.select_input_file)
        self.btn_output.clicked.connect(self.select_output_file)
        
        # åˆå§‹åŒ–ï¼šæ‰‹åŠ¨è§¦å‘ä¸€æ¬¡èŒƒå›´è®¾å®šï¼ˆé»˜è®¤ CQPï¼‰
        self.update_slider_range("cqp")
        
    
    def update_slider_range(self, mode):
        """æ ¹æ®é€‰æ‹©çš„ RC æ¨¡å¼ï¼ŒåŠ¨æ€è°ƒæ•´æ»‘å—çš„æœ€å°å€¼ã€æœ€å¤§å€¼å’Œå½“å‰å€¼"""
        # æš‚æ—¶é™éŸ³ä¿¡å·ï¼Œé˜²æ­¢åœ¨åˆ‡èŒƒå›´æ—¶è§¦å‘å¤šä½™çš„ label æ›´æ–°å¯¼è‡´æŠ¥é”™
        self.sld_v_value.blockSignals(True)
        
        if mode == "cqp":
            # CQP ä¿æŒçº¿æ€§ï¼Œ0-51 è¶Šå°ç”»è´¨è¶Šå¥½
            self.sld_v_value.setRange(0, 51)
            self.sld_v_value.setSingleStep(1)
            self.sld_v_value.setValue(32) 
        else:
            # CBR/VBR æ¨¡å¼ï¼šç‰©ç†æ»‘å—å˜æˆ 0-100 çš„è¿›åº¦æ¡
            self.sld_v_value.setRange(0, 100)
            self.sld_v_value.setSingleStep(1)
            # é»˜è®¤ç»™ä¸ª 54 å·¦å³çš„ä½ç½®ï¼Œå› ä¸º 54^3 æ˜ å°„å‡ºæ¥åˆšå¥½å¤§æ¦‚æ˜¯ 5000kbps å·¦å³
            self.sld_v_value.setValue(56) 
            
        self.sld_v_value.blockSignals(False)
        self.update_slider_label()
    
    def get_mapped_bitrate(self, slider_val):
        """å°† 0-100 çš„æ»‘å—å€¼è¿›è¡Œéçº¿æ€§(ä¸‰æ¬¡å¹‚)æ˜ å°„åˆ° 50-30000 kbps"""
        min_kbps = 50
        max_kbps = 30000
        # ä½¿ç”¨ä¸‰æ¬¡å¹‚å‡½æ•°ï¼šæ»‘å—åœ¨å‰åŠæ®µæ•°å­—å˜åŒ–ææ…¢ï¼ŒååŠæ®µå˜åŒ–æå¿«
        ratio = slider_val / 100.0
        mapped_val = min_kbps + (max_kbps - min_kbps) * (ratio ** 3)
        # æŠŠè®¡ç®—ç»“æœè§„æ•´ä¸€ä¸‹ï¼Œå‘ä¸‹å–æ•´åˆ°æœ€æ¥è¿‘çš„ 10ï¼Œè®© UI çœ‹èµ·æ¥æ›´æ•´æ´
        return int(round(mapped_val / 10) * 10)
    
    def get_reverse_mapped_slider_val(self, target_kbps):
        """å°†çœŸå®çš„ç ç‡ (å¦‚ 5000 kbps) é€†å‘æ¨å¯¼å› 0-100 çš„æ»‘å—ç‰©ç†åˆ»åº¦"""
        min_kbps = 50
        max_kbps = 30000
        if target_kbps <= min_kbps: return 0
        if target_kbps >= max_kbps: return 100
        # é€†å‘å…¬å¼ï¼šå¼€ä¸‰æ¬¡æ–¹æ ¹
        ratio = ((target_kbps - min_kbps) / (max_kbps - min_kbps)) ** (1/3.0)
        return int(round(ratio * 100))
    
    def update_slider_label(self):
        """å®æ—¶æ›´æ–°æ»‘å—æ—è¾¹çš„æ–‡å­—æ˜¾ç¤º"""
        mode = self.cb_v_rc.currentText()
        if mode == "cqp":
            val = self.sld_v_value.value()
            self.lbl_v_val_display.setText(str(val))
        else:
            # VBR/CBR æ¨¡å¼ä¸‹ï¼Œæ–‡å­—æ˜¾ç¤ºçš„æ˜¯ç»è¿‡æ•°å­¦æ˜ å°„åçš„çœŸå®ç ç‡
            val = self.get_mapped_bitrate(self.sld_v_value.value())
            if val >= 1000:
                self.lbl_v_val_display.setText(f"{val/1000:.1f} Mbps")
            else:
                self.lbl_v_val_display.setText(f"{val} kbps")
    
    def probe_hardware_encoders(self):
        print("æ­£åœ¨è¿›è¡Œå…¨ç¡¬ä»¶å¼•æ“ç‚¹ç«è‡ªæ£€...")
        test_encoders = [
            "av1_nvenc", "hevc_nvenc", "h264_nvenc", # NVIDIA æå‰ï¼ˆå› ä¸ºä½ æ˜¯ N å¡ç”¨æˆ·ï¼‰
            "av1_amf", "hevc_amf", "h264_amf",       # AMD
            "av1_qsv", "hevc_qsv", "h264_qsv",       # Intel
            "libsvtav1", "libx265", "libx264"        # CPU
        ]
        
        available = []
        CREATE_NO_WINDOW = 0x08000000
        
        progress = QProgressDialog("æ­£åœ¨åˆå§‹åŒ–ç¡¬ä»¶æ¢é’ˆ...", "è·³è¿‡", 0, len(test_encoders), self)
        progress.setWindowTitle("å¼•æ“è‡ªæ£€")
        progress.setWindowModality(Qt.WindowModal)
        progress.show()
        
        for i, enc in enumerate(test_encoders):
            progress.setLabelText(f"æ­£åœ¨æµ‹è¯• {enc} å¼•æ“...")
            progress.setValue(i)
            QApplication.processEvents() 
            if progress.wasCanceled(): break

            # === NVIDIA å…¼å®¹æ€§è¡¥ä¸æŒ‡ä»¤ ===
            cmd = [
                get_ext_path("ffmpeg.exe"), "-y", 
                "-f", "lavfi", "-i", "color=c=black:s=320x240", # å¢å¤§åˆ†è¾¨ç‡ï¼Œé¿å¼€å¯¹é½é™åˆ¶
                "-vframes", "1", 
                "-c:v", enc, 
                "-pix_fmt", "yuv420p", # ï¼ï¼ï¼æ ¸å¿ƒï¼šå¼ºåˆ¶æŒ‡å®š NVENC æœ€å–œæ¬¢çš„ yuv420p æ ¼å¼
                "-f", "null", "-"
            ]
            
            try:
                # === æ ¸å¿ƒæ”¹è¿›ï¼šè¶…æ—¶å¢åŠ åˆ° 5 ç§’ï¼Œç»™ N å¡ CUDA åˆå§‹åŒ–ç•™è¶³æ—¶é—´ ===
                result = subprocess.run(
                    cmd, 
                    stdout=subprocess.PIPE, 
                    stderr=subprocess.PIPE, 
                    text=True, 
                    encoding='utf-8', 
                    creationflags=CREATE_NO_WINDOW, 
                    timeout=5 
                )
                
                if result.returncode == 0:
                    available.append(enc)
                    print(f"âœ… æ¢æµ‹æˆåŠŸ: {enc}")
                else:
                    # å³ä½¿å¤±è´¥ï¼Œæˆ‘ä»¬ä¹Ÿè¦çœ‹ä¸€çœ¼ä¸ºä»€ä¹ˆå¤±è´¥ (ç‰¹åˆ«æ˜¯ N å¡)
                    print(f"âŒ {enc} å¤±è´¥åŸå› æ‘˜è¦: {result.stderr[-100:]}")
            except Exception as e:
                print(f"âš ï¸ {enc} æ¢æµ‹è¶…æ—¶æˆ–å¼‚å¸¸: {e}")
                
        progress.setValue(len(test_encoders)) 
        # ç”Ÿæˆç¡¬ä»¶è‡ªæ£€æŠ¥å‘Šï¼Œæ–¹ä¾¿ä½ åœ¨åˆ«äººç”µè„‘ä¸Šæ’æŸ¥
        with open("hardware_report.txt", "w", encoding="utf-8") as f:
            f.write(f"å¯ç”¨ç¼–ç å™¨åˆ—è¡¨: {available}\n")
            f.write(f"FFmpeg è·¯å¾„: {get_ext_path('ffmpeg.exe')}\n")
        available.append("copy") 
        return available
    
    def load_dynamic_presets(self):
        import yaml # å±€éƒ¨å¼•å…¥ï¼Œä¿æŒé¡¶éƒ¨æ•´æ´
        
        self.preset_configs = {}
        raw_presets = []

        # 1. åŠ¨æ€å®šä½ config/presets.yaml çš„ç»å¯¹è·¯å¾„
        yaml_path = os.path.join(get_app_dir(), "config", "presets.yaml")

        # 2. å®‰å…¨è¯»å– YAML æ–‡ä»¶
        try:
            with open(yaml_path, 'r', encoding='utf-8') as f:
                data = yaml.safe_load(f)
                raw_presets = data.get("presets", [])
            print(f"ğŸ“„ æˆåŠŸåŠ è½½å¤–éƒ¨é¢„è®¾é…ç½®ï¼Œå…±è¯»å–åˆ° {len(raw_presets)} ä¸ªé¢„è®¾ã€‚")
        except FileNotFoundError:
            print(f"âš ï¸ æ‰¾ä¸åˆ°é…ç½®æ–‡ä»¶: {yaml_path}ï¼Œè¯·æ£€æŸ¥ config ç›®å½•ï¼")
        except Exception as e:
            print(f"ğŸ’¥ è¯»å– presets.yaml å‘ç”Ÿè¯­æ³•é”™è¯¯: {e}")

        # ==========================================
        # 3. é€å­—ä¿ç•™ï¼šåŸæ±åŸå‘³çš„æ ¸å¿ƒåŒ¹é…ä¸æ³¨å…¥é€»è¾‘
        # ==========================================
        for p in raw_presets:
            matched_encoder = next((enc for enc in self.available_v_encoders if p["requires"] in enc), None)
            
            if matched_encoder:
                config = p["ui_state"].copy()
                config["v_enc"] = matched_encoder # åŠ¨æ€å¡å…¥å¯ç”¨çš„ç¡¬ä»¶ç¼–ç å™¨
                self.preset_configs[p["name"]] = config

        # æ°¸è¿œåœ¨åˆ—è¡¨æœ€åä¿ç•™â€œè‡ªå®šä¹‰â€é€‰é¡¹
        self.preset_configs["âš™ï¸ è‡ªå®šä¹‰å‚æ•°..."] = {}
        
    def load_tooltips(self):
        import yaml
        import os
        
        yaml_path = os.path.join(get_app_dir(), "config", "tooltips.yaml")
        try:
            with open(yaml_path, 'r', encoding='utf-8') as f:
                tips = yaml.safe_load(f)
            
            # ä» YAML å­—å…¸ä¸­æå–å¯¹åº”æ¿å—å¹¶æ³¨å…¥ UI
            if "encoder_tips" in tips:
                self.set_combo_tooltips(self.cb_v_encoder, tips["encoder_tips"])
            if "preset_tips" in tips:
                self.set_combo_tooltips(self.combo_preset, tips["preset_tips"])
            if "rc_tips" in tips:
                self.set_combo_tooltips(self.cb_v_rc, tips["rc_tips"])
                
            print("ğŸ’¬ æˆåŠŸåŠ è½½å¤–éƒ¨æ‚¬æµ®ç§‘æ™®æç¤ºåº“ã€‚")
        except FileNotFoundError:
            print(f"âš ï¸ æ‰¾ä¸åˆ°æç¤ºæ–‡æ¡ˆåº“: {yaml_path}")
        except Exception as e:
            print(f"ğŸ’¥ è¯»å– tooltips.yaml å¤±è´¥: {e}")

    def toggle_custom_tab(self, text):
        try:
            #print(f"\nğŸ”„ æ”¶åˆ°é¢„è®¾åˆ‡æ¢è¯·æ±‚: {text}")
            if "è‡ªå®šä¹‰" in text:
                self.tab_custom.setVisible(True)
            else:
                self.tab_custom.setVisible(False)
                
            if text not in self.preset_configs or not self.preset_configs[text]:
                #print("âš ï¸ è¯¥é€‰é¡¹ä¸ºè‡ªå®šä¹‰æˆ–ç©ºé¢„è®¾ï¼Œè·³è¿‡è‡ªåŠ¨æ‹¨åŠ¨ã€‚")
                return
                
            cfg = self.preset_configs[text]
            #print(f"âœ… æå–åˆ°é¢„è®¾é…ç½®: {cfg}")

            # æš‚æ—¶å±è”½ä¿¡å·
            self.cb_v_encoder.blockSignals(True)
            self.cb_v_fps.blockSignals(True)
            self.cb_v_res.blockSignals(True)
            self.cb_v_rc.blockSignals(True)
            self.sld_v_value.blockSignals(True)
            self.cb_a_encoder.blockSignals(True)
            self.cb_a_bitrate.blockSignals(True)
            self.cb_a_sample.blockSignals(True)

            #print("ğŸ‘‰ æ­£åœ¨æ‹¨åŠ¨è§†é¢‘åŸºç¡€å‚æ•°...")
            self.cb_v_encoder.setCurrentText(cfg["v_enc"])
            self.cb_v_fps.setCurrentText(cfg["fps"])
            self.cb_v_res.setCurrentText(cfg["res"])
            
            #print("ğŸ‘‰ æ­£åœ¨æ‹¨åŠ¨ç ç‡æ§åˆ¶æ¨¡å¼...")
            self.cb_v_rc.setCurrentText(cfg["rc"])
            self.cb_v_rc.blockSignals(False)
            self.update_slider_range(cfg["rc"]) 
            
            #print("ğŸ‘‰ æ­£åœ¨è®¡ç®—å¹¶æ‹¨åŠ¨æ»‘å—...")
            self.sld_v_value.blockSignals(True) 
            if cfg["rc"] == "cqp":
                self.sld_v_value.setValue(cfg["val"])
            else:
                slider_pos = self.get_reverse_mapped_slider_val(cfg["val"])
                self.sld_v_value.setValue(slider_pos)
                
            #print("ğŸ‘‰ æ­£åœ¨æ‹¨åŠ¨éŸ³é¢‘å‚æ•°...")
            self.cb_a_encoder.setCurrentText(cfg["a_enc"])
            self.cb_a_bitrate.setCurrentText(cfg["a_bit"])
            self.cb_a_sample.setCurrentText(cfg["a_sample"])

            #print("âœ… æ‹¨åŠ¨å®Œæˆï¼Œæ­£åœ¨æ¢å¤ä¿¡å·ï¼")
            self.cb_v_encoder.blockSignals(False)
            self.cb_v_fps.blockSignals(False)
            self.cb_v_res.blockSignals(False)
            self.sld_v_value.blockSignals(False)
            self.cb_a_encoder.blockSignals(False)
            self.cb_a_bitrate.blockSignals(False)
            self.cb_a_sample.blockSignals(False)
            
            self.update_slider_label()
            #print("ğŸ‰ UI é¢„è®¾çŠ¶æ€åŒæ­¥å®Œç¾ç»“æŸï¼\n")

        except Exception as e:
            import traceback
            print(f"\nğŸ’¥ğŸ’¥ åˆ‡æ¢é¢„è®¾æ—¶å‘ç”Ÿè‡´å‘½é”™è¯¯: {e}")
            traceback.print_exc()
            print("ğŸ’¥ğŸ’¥ ä¸Šé¢çš„æŠ¥é”™å°±æ˜¯å¯¼è‡´å‚æ•°æ²¡å˜çš„ç½ªé­ç¥¸é¦–ï¼\n")
    
    def change_output_extension(self, new_ext):
        # æ‹¿åˆ°å½“å‰è¾“å…¥æ¡†é‡Œçš„å®Œæ•´è·¯å¾„
        current_path = self.txt_output.text().strip()
        
        # å¦‚æœè·¯å¾„æ˜¯ç©ºçš„ï¼ˆä¹Ÿå°±æ˜¯ç”¨æˆ·è¿˜æ²¡å¯¼å…¥è§†é¢‘ï¼‰ï¼Œé‚£å°±ä»€ä¹ˆéƒ½ä¸åš
        if not current_path:
            return
            
        import os
        # ç¥å¥‡çš„ os.path.splitextï¼šç›´æ¥æŠŠè·¯å¾„åŠˆæˆâ€œæ— åç¼€çš„çº¯è·¯å¾„â€å’Œâ€œæ—§åç¼€â€ä¸¤åŠ
        base_path, old_ext = os.path.splitext(current_path)
        
        # é‡æ–°æ‹¼æ¥ä¸Šåˆšåˆšé€‰ä¸­çš„æ–°åç¼€ï¼Œå¹¶å¡å›è¾“å…¥æ¡†
        new_path = base_path + new_ext
        self.txt_output.setText(new_path)
    
    def build_ffmpeg_args(self):
        # âœ¨ æ ¸å¿ƒä¿®å¤ 3ï¼šå½»åº•åˆ é™¤äº†æœ€å‰é¢çš„ if æ‹¦æˆªå™¨ï¼
        # ç°åœ¨çš„å”¯ä¸€çœŸç†æ¥æºå°±æ˜¯ UI é¢æ¿ï¼
        args = []
        v_enc = self.cb_v_encoder.currentText()
        is_nvenc = "nvenc" in v_enc
        is_amf = "amf" in v_enc
        is_qsv = "qsv" in v_enc
        is_cpu = "lib" in v_enc

        # --- è§†é¢‘ç¼–ç éƒ¨åˆ† ---
        if v_enc == "copy":
            args.extend(["-c:v", "copy"])
        else:
            args.extend(["-c:v", v_enc])
            
            # NVIDIA / AMD ç¡¬ä»¶ H264 çš„æŠ¤åŸæ²³ (ä¿®å¤æš´æ¯™é—®é¢˜)
            if v_enc in ["h264_nvenc", "h264_amf"]:
                args.extend(["-pix_fmt", "yuv420p", "-profile:v", "high"])
            
            # å¸§ç‡å¤„ç†
            fps = self.cb_v_fps.currentText()
            if fps != "ä¿æŒæº": 
                args.extend(["-r", fps])
            
            # åˆ†è¾¨ç‡å¤„ç†
            res = self.cb_v_res.currentText()
            if res == "1080p": args.extend(["-vf", "scale=-1:1080"])
            elif res == "720p": args.extend(["-vf", "scale=-1:720"])
            elif res == "2160p": args.extend(["-vf", "scale=-1:2160"])
            elif res == "1440p": args.extend(["-vf", "scale=-1:1440"])
                
            # --- ç ç‡æ§åˆ¶é€‚é… (æ¥å…¥æœ€æ–°éçº¿æ€§æ»‘å—é€»è¾‘) ---
            rc_mode = self.cb_v_rc.currentText()
            
            if rc_mode == "cqp":
                val = str(self.sld_v_value.value()) # CQP ä¿æŒçº¿æ€§ï¼Œç›´æ¥æ‹¿åŸå€¼
                if is_nvenc:
                    # NVENC çœŸæ­£çš„â€œæ’å®šç”»è´¨â€æœ€ä½³å®è·µ
                    args.extend(["-rc", "vbr", "-cq", val, "-b:v", "0"])
                elif is_amf:
                    args.extend(["-rc", "cqp", "-qp_i", val, "-qp_p", val])
                elif is_qsv:
                    args.extend(["-global_quality", val])
                else:
                    args.extend(["-crf", val])
                    
            elif rc_mode == "vbr":
                val_int = self.get_mapped_bitrate(self.sld_v_value.value()) # VBR æ‹¿éçº¿æ€§æ˜ å°„åçš„çœŸå®ç ç‡
                val = f"{val_int}k"
                if is_nvenc:
                    args.extend(["-rc", "vbr", "-b:v", val, "-maxrate:v", val, "-bufsize:v", val])
                elif is_amf:
                    args.extend(["-rc", "vbr_peak", "-b:v", val])
                else:
                    args.extend(["-b:v", val])
                    
            elif rc_mode == "cbr":
                val_int = self.get_mapped_bitrate(self.sld_v_value.value()) # CBR æ‹¿éçº¿æ€§æ˜ å°„åçš„çœŸå®ç ç‡
                val = f"{val_int}k"
                if is_nvenc:
                    args.extend(["-rc", "cbr", "-b:v", val, "-maxrate:v", val, "-bufsize:v", val])
                elif is_amf:
                    args.extend(["-rc", "cbr", "-b:v", val])
                else:
                    args.extend(["-b:v", val, "-maxrate:v", val, "-bufsize:v", val])

        # --- éŸ³é¢‘éƒ¨åˆ† ---
        a_enc = self.cb_a_encoder.currentText()
        if "å‰¥ç¦»é™éŸ³" in a_enc: 
            args.extend(["-an"])
        elif a_enc == "copy": 
            args.extend(["-c:a", "copy"])
        else:
            args.extend(["-c:a", a_enc])
            ab = self.cb_a_bitrate.currentText()
            args.extend(["-b:a", ab])
            ar = self.cb_a_sample.currentText()
            if ar != "ä¿æŒæº": 
                args.extend(["-ar", ar])

        return args
    
    def start_encoding(self):
        # 1. åŠ¨æ€è·å–ç•Œé¢ä¸Šè¾“å…¥æ¡†é‡Œçš„è·¯å¾„
        input_path = self.txt_input.text().strip()
        output_path = self.txt_output.text().strip()

        # ç®€å•æ‹¦æˆªï¼šå¦‚æœæ²¡é€‰æ–‡ä»¶å°±ç‚¹å¼€å§‹ï¼Œå¼¹çª—è­¦å‘Šå¹¶æ‰“æ–­
        if not input_path or not output_path:
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©éœ€è¦å‹åˆ¶çš„è§†é¢‘ï¼")
            return

        # 2. é”å®šæŒ‰é’®ï¼Œé˜²æ­¢é‡å¤ç‚¹å‡»ï¼Œæ¿€æ´»æš‚åœå’Œåœæ­¢æŒ‰é’®
        self.btn_start.setEnabled(False)
        self.btn_pause.setEnabled(True)
        self.btn_stop.setEnabled(True)
        self.btn_start.setText("â³ å‹åˆ¶ä¸­...")
        self.lbl_status.setText("çŠ¶æ€: è¿æ¥ç¼–ç å™¨...")

        # 3. åœ¨å¼€å¯å¤šçº¿ç¨‹å‰ï¼Œç”¨æ¢é’ˆç¬é—´è¯»å–çœŸå®æ—¶é•¿å¹¶ä¿å­˜ä¸ºå®ä¾‹å±æ€§
        self.total_seconds = get_video_duration(input_path)
        print(f"æ¢é’ˆæˆåŠŸè·å–è§†é¢‘æ€»æ—¶é•¿: {self.total_seconds} ç§’")

        # 4. === æ‰¾å›å¤±è¸ªçš„é¢„è§ˆå¼€å…³é€»è¾‘ä¸è·¯å¾„ç”Ÿæˆ ===
        self.enable_preview = self.chk_preview.isChecked()
        self.preview_temp_dir = tempfile.gettempdir() 
        self.preview_path = os.path.join(self.preview_temp_dir, "ffmpeg_preview_temp.jpg")

        # å¦‚æœä¸Šæ¬¡çš„æ®‹ç•™å›¾ç‰‡è¿˜åœ¨ï¼Œå…ˆæ¸…ç†æ‰é¿å…ç”»é¢ç©¿è¶Š
        if os.path.exists(self.preview_path):
            try:
                os.remove(self.preview_path)
            except:
                pass

        # å¯åŠ¨å‰å°ç›‘è§†å™¨å®šæ—¶å™¨
        self.preview_timer = QTimer(self)
        self.preview_timer.timeout.connect(self.update_preview)
        if self.enable_preview:
            self.lbl_preview.setText("æ­£åœ¨è¿æ¥ç”»é¢æº...")
            self.preview_timer.start(1000) 

        # 5. å‘¼å«ç¿»è¯‘å®˜ï¼ŒæŠŠå½“å‰ç•Œé¢çš„é€‰æ‹©ç¿»è¯‘æˆå‚æ•°åˆ—è¡¨
        dynamic_args = self.build_ffmpeg_args()
        print(f"ç”Ÿæˆçš„å‹åˆ¶å‚æ•°: {dynamic_args}") 

        # 6. åˆ›å»ºå¹¶å¯åŠ¨åå°å¤§å¿ƒè„ (ä¼ å…¥åŠ¨æ€è·å–çš„è·¯å¾„ã€å¼€å…³çŠ¶æ€å’Œç¿»è¯‘å¥½çš„å‚æ•°)
        self.worker = FFmpegWorker(input_path, output_path, self.enable_preview, self.preview_path, dynamic_args)
        self.worker.log_signal.connect(self.print_log)
        self.worker.finished_signal.connect(self.encoding_finished)
        self.worker.start()
        
    def toggle_pause(self):
        if self.btn_pause.text() == "â¸ æš‚åœ":
            self.worker.pause()
            self.btn_pause.setText("â–¶ æ¢å¤")
            self.lbl_status.setText("çŠ¶æ€: å·²æš‚åœ (æ˜¾å¡å·²æŒ‚èµ·)")
            if hasattr(self, 'preview_timer'): self.preview_timer.stop()
        else:
            self.worker.resume()
            self.btn_pause.setText("â¸ æš‚åœ")
            self.lbl_status.setText("çŠ¶æ€: æ­£åœ¨ç‹‚é£™å‹åˆ¶ä¸­...")
            if hasattr(self, 'preview_timer'): self.preview_timer.start(1000)
            
    def stop_encoding(self):
        # åªè¦è§¦å‘åœæ­¢æŒ‰é’®ï¼Œåªç®¡æ€åå°ï¼Œåç»­çš„ UI æ›´æ–°å…¨éƒ¨äº¤ç»™ä¿¡å·è‡ªç„¶è§¦å‘
        if hasattr(self, 'worker') and self.worker.isRunning():
            self.worker.stop()

    def print_log(self, text):
        #print(text) #è°ƒè¯•æ—¶ç›´æ¥å¾€æ§åˆ¶å°è¾“å‡º
        # 1. ä½¿ç”¨æ­£åˆ™è¡¨è¾¾å¼ç‹™å‡»â€œå½“å‰æ—¶é—´â€å’Œâ€œå‹åˆ¶é€Ÿåº¦â€
        # åŒ¹é…æ ¼å¼å¦‚ time=01:14:58.85
        time_match = re.search(r"time=(\d{2}:\d{2}:\d{2}\.\d{2})", text)
        # åŒ¹é…æ ¼å¼å¦‚ speed=12.9x
        speed_match = re.search(r"speed=\s*([\d\.]+)x", text)
        storage_match = re.search(r"size=\s*([\d\.]+)KiB", text)

        if time_match:
            time_str = time_match.group(1)
            
            # 2. å°† 01:14:58.85 æ¢ç®—æˆçº¯ç§’æ•°
            h, m, s = time_str.split(':')
            current_seconds = int(h) * 3600 + int(m) * 60 + float(s)

            # 3 & 4. === ä¿®æ”¹ï¼šä½¿ç”¨æ¢é’ˆè¯»å–åˆ°çš„çœŸå®æ—¶é•¿æ¥è®¡ç®—ç™¾åˆ†æ¯” ===
            percent = int((current_seconds / self.total_seconds) * 100)
            
            # é™åˆ¶åœ¨ 0-100 ä¹‹é—´ï¼Œé˜²æ­¢æµ®ç‚¹å¾®å°è¯¯å·®å¯¼è‡´è¿›åº¦æ¡æº¢å‡ºæŠ¥é”™
            percent = max(0, min(100, percent))
            self.progress_bar.setValue(percent)
            
            # æ ¼å¼åŒ–å‚¨å­˜å¤§å°æ˜¾ç¤º
            if storage_match:
                size_kib = float(storage_match.group(1))
                if size_kib >= 1000 * 1000:
                    storage_text = f"{size_kib / (1000 * 1000):.2f} GiB"
                elif size_kib >= 1000:
                    storage_text = f"{size_kib / 1000:.2f} MiB"
                else:
                    storage_text = f"{size_kib:.2f} KiB"

            # 5. æ›´æ–°çŠ¶æ€æ é¢æ¿
            speed_text = speed_match.group(1) if speed_match else "--"
            self.lbl_status.setText(f"çŠ¶æ€: ç‹‚é£™å‹åˆ¶ä¸­... | é€Ÿåº¦: {speed_text}x | å½“å‰è¿›åº¦: {time_str} | å½“å‰æ–‡ä»¶å¤§å°ï¼š{storage_text}")
        
    def update_preview(self):
        if not os.path.exists(self.preview_path):
            return

        try:
            # 1. ç¬é—´å°†ç¡¬ç›˜æ–‡ä»¶çš„æ‰€æœ‰å­—èŠ‚å¸å…¥ Python å†…å­˜
            with open(self.preview_path, 'rb') as f:
                data = f.read()
            
            # 2. åŸºç¡€è¿‡æ»¤ï¼šå¦‚æœå›¾ç‰‡è¿ 1KB éƒ½æ²¡æœ‰ï¼Œè¯´æ˜ FFmpeg åˆšæ¸…ç©ºæ–‡ä»¶ï¼Œç›´æ¥è·³è¿‡
            if len(data) < 1024:
                return

            # 3. ä¸¥æ ¼æ ¡éªŒå†…å­˜æ•°æ®ï¼šå¿…é¡»ä»¥ FF D8 å¼€å¤´ (SOF)ï¼Œä¸”ä»¥ FF D9 ç»“å°¾ (EOF)
            if not (data.startswith(b'\xff\xd8') and data.endswith(b'\xff\xd9')):
                return

            # 4. æ ¸å¿ƒé­”æ³•ï¼šåˆ‡æ–­ç¡¬ç›˜è”ç³»ï¼ç›´æ¥è®© Qt ä»åˆšæ‰å¸å…¥çš„å†…å­˜å­—èŠ‚ (data) é‡Œè¯»å–ç”»é¢
            pixmap = QPixmap()
            if pixmap.loadFromData(data):
                scaled_pixmap = pixmap.scaled(
                    self.lbl_preview.size(), 
                    Qt.KeepAspectRatio, 
                    Qt.SmoothTransformation
                )
                self.lbl_preview.setPixmap(scaled_pixmap)

        except Exception:
            # å…œåº•ï¼šå¦‚æœè¿™ 1 æ¯«ç§’åˆšå¥½ç¢°åˆ°ç³»ç»Ÿçº§çš„æ–‡ä»¶æ­»é”ï¼Œç›´æ¥é™é»˜æ”¾è¿‡ï¼Œç­‰ä¸‹ä¸€ç§’
            pass

    def encoding_finished(self):
        # 1. æ¢å¤æ‰€æœ‰æŒ‰é’®çš„åŸºç¡€çŠ¶æ€
        self.btn_start.setEnabled(True)
        self.btn_start.setText("ğŸš€ å¼€å§‹") # === æ–°å¢ï¼šå°†æŒ‰é’®æ–‡å­—å½»åº•æ¢å¤åˆå§‹çŠ¶æ€ ===
        self.btn_pause.setEnabled(False)
        self.btn_stop.setEnabled(False)
        self.btn_pause.setText("â¸ æš‚åœ")
        
        # 2. åœæ­¢å¯èƒ½è¿˜åœ¨åˆ·æ–°çš„ç›‘è§†å™¨å®šæ—¶å™¨
        if hasattr(self, 'preview_timer') and self.preview_timer.isActive():
            self.preview_timer.stop()

        # 3. æ ¸å¿ƒåˆ†æµï¼šåˆ¤æ–­åˆ°åº•æ˜¯æ­£å¸¸è·‘å®Œï¼Œè¿˜æ˜¯è¢«ä¸­é€”å¹²æ‰çš„ï¼Ÿ
        if getattr(self.worker, 'is_cancelled', False):
            # è¢«å¼ºè¡Œä¸­æ­¢çš„ UI é€»è¾‘
            self.lbl_status.setText("çŠ¶æ€: å‹åˆ¶å·²å¼ºåˆ¶ä¸­æ­¢ ğŸš«")
            if self.enable_preview:
                self.lbl_preview.clear()
                self.lbl_preview.setText("ä»»åŠ¡å·²å–æ¶ˆ\n(ç”»é¢é¢„è§ˆç»“æŸ)")
            print("====== å‹åˆ¶å·²è¢«ç”¨æˆ·ä¸­æ­¢ï¼======")
        else:
            # æ­£å¸¸é¡ºåˆ©å®Œæˆçš„ UI é€»è¾‘
            self.lbl_status.setText("çŠ¶æ€: å‹åˆ¶å®Œæˆï¼ âœ…")
            self.progress_bar.setValue(100) # åªæœ‰æ­£å¸¸å®Œæˆæ‰å¼ºè¡Œæ‹‰æ»¡è¿›åº¦æ¡
            if self.enable_preview:
                self.lbl_preview.clear()
                self.lbl_preview.setText("å‹åˆ¶å·²å®Œæˆ\n(ç”»é¢é¢„è§ˆç»“æŸ)")
            print("====== å‹åˆ¶å½»åº•ç»“æŸï¼======")
    
    def select_input_file(self):
        # å‘¼å‡º Windows åŸç”Ÿæ–‡ä»¶é€‰æ‹©æ¡†ï¼Œé™åˆ¶åªèƒ½é€‰å¸¸è§è§†é¢‘æ ¼å¼
        file_path, _ = QFileDialog.getOpenFileName(self, "é€‰æ‹©åŸè§†é¢‘", "", "è§†é¢‘æ–‡ä»¶ (*.mp4 *.mkv *.mov *.avi *.mkv);;æ‰€æœ‰æ–‡ä»¶ (*.*)")
        if file_path:
            # æŠŠé€‰ä¸­çš„è·¯å¾„å¡«å…¥è¾“å…¥æ¡†
            self.txt_input.setText(file_path)
            
            # æ¯”å¦‚è¾“å…¥æ˜¯ D:/video.mp4ï¼Œè¾“å‡ºè‡ªåŠ¨å˜æˆ D:/video_output.mp4
            default_out = file_path.rsplit('.', 1)[0] + "_output.mp4"
            self.txt_output.setText(default_out)
        
        if file_path:
            # æ¢ä¸Šä¸€å¥—æå®¢ä¸“ç”¨çš„è§å…‰é’è‰²ã€ç­‰å®½å­—ä½“æ ·å¼
            self.lbl_preview.setStyleSheet("background-color: #0b0c10; color: #45a3ad; border-radius: 8px; font-weight: bold; font-family: Consolas, monospace; font-size: 16px;")
            self.lbl_preview.setText("æ­£åœ¨æ‰«æè§†é¢‘åº•å±‚æ•°æ®...")
            
            from PySide6.QtWidgets import QApplication
            QApplication.processEvents() # å¼ºåˆ¶åˆ·æ–° UIï¼Œè®©æ–‡å­—ç¬é—´äº®èµ·
            
            # å‘¼å«æ¢é’ˆï¼ŒæŠŠæ‹¿åˆ°æƒ…æŠ¥è´´åœ¨å±å¹•ä¸Š
            info_text = probe_video_info(file_path)
            self.lbl_preview.setText(info_text)

    def select_output_file(self):
        # å‘¼å‡º Windows åŸç”Ÿä¿å­˜æ¡†
        file_path, _ = QFileDialog.getSaveFileName(self, "è®¾ç½®å¯¼å‡ºè·¯å¾„", self.txt_output.text(), "è§†é¢‘æ–‡ä»¶ (*.mp4)")
        if file_path:
            self.txt_output.setText(file_path)

    def closeEvent(self, event: QCloseEvent):
        if hasattr(self, 'worker') and self.worker.isRunning():
            reply = QMessageBox.question(self, 'ç¡®è®¤é€€å‡º', "å‹åˆ¶å°šæœªå®Œæˆï¼Œç¡®å®šè¦å¼ºè¡Œé€€å‡ºå¹¶æ”¾å¼ƒä»»åŠ¡å—ï¼Ÿ",
                                        QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
            if reply == QMessageBox.Yes:
                self.worker.stop()
                event.accept()
            else:
                event.ignore()
        else:
            event.accept()

    def set_combo_tooltips(self, combo, tooltips_dict):
        """
        ä¸º QComboBox çš„æ¯ä¸ªé€‰é¡¹è®¾ç½®æ‚¬æµ®è¯´æ˜
        :param combo: ç›®æ ‡ä¸‹æ‹‰èœå•å¯¹è±¡
        :param tooltips_dict: æ ¼å¼ä¸º {"é€‰é¡¹å": "è¯´æ˜æ–‡å­—"} çš„å­—å…¸
        """
        model = combo.model() # è·å–ä¸‹æ‹‰èœå•èƒŒåçš„æ•°æ®æ¨¡å‹
        for i in range(combo.count()):
            text = combo.itemText(i)
            if text in tooltips_dict:
                # æ ¸å¿ƒï¼šå°†è¯´æ˜æ–‡å­—æ³¨å…¥åˆ°è¯¥é€‰é¡¹çš„ ToolTip è§’è‰²ä¸­
                model.setData(model.index(i, 0), tooltips_dict[text], Qt.ToolTipRole)
    

if __name__ == "__main__":
    init_config_files()
    app = QApplication(sys.argv)
    window = FFmpegGUI()
    window.show()
    sys.exit(app.exec())